### PR TITLE
patch for 6068, adding a flag about clearing logger on session end

### PR DESCRIPTION
**Pull request by wyhmichael**
## Proposed changes

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

This is patch for #6068, and workaround for issue mentioned in https://github.com/webdriverio/webdriverio/issues/6346#issuecomment-1179299272, by adding a flag to allow turn on/off logger cleanup on session end

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### Reviewers: @webdriverio/project-committers

<a href="https://gitpod.io/#https://github.com/webdriverio/webdriverio/pull/8506"></a>

### CODE REVIEW

**Comment by christian-bromann** [comment]
Adding an option to provide a workaround is not an ideal solution. Can we analyse what the use cases are here and maybe find a solution that doesn't require to add this option?

**Comment by wyhmichael** [comment]
Hi @christian-bromann sorry for the delay, I made a sample repo to show the use cases of wido logger - https://github.com/wyhmichael/wdio-logger-afterSession-issue

There're 3 tests: 

1. loggerOnAfterSessionHook.e2e.ts (runner mode)
2. loggerOnSessionReload.e2e.ts (runner mode)
3. sequentialWDIOSessions.e2e.js (standalone mode) : This test is meant to simulate the issue mentioned in https://github.com/webdriverio/webdriverio/issues/6068

With the introduction of `clearLogger` in https://github.com/webdriverio/webdriverio/issues/6068, the use case 1 and 2 above are behaving like how it was before where all the logs would be written and kept in the log file, now the log file would be cleared whenever `afterSession` and `reloadSession` called (which trigger `deleteSession` command to trigger `clearLogger`)

**Comment by christian-bromann** [comment]
How about we disable the `clearLogger` function when the testrunner is running, e.g. `WDIO_LOG_PATH` is set as environment variable. This would still allow the use case for what it got implemented and would fix the bug you are experiencing.

**Comment by christian-bromann** [comment]
It would be:
```js
if (!process.env.WDIO_LOG_PATH) {
  logger.clearLogger();
}
```

but yes!

**Comment by wyhmichael** [comment]
Thanks @christian-bromann! 

I tried to update [this condition](https://github.com/webdriverio/webdriverio/pull/8506/files#diff-b08c37e616f7015c3c66a75fde5d6a5f9e58955dcc8be341320d65536e492037R104) to be 

```
if (command === 'deleteSession' && !process.env.WDIO_LOG_PATH) {
```

but looks like [`process.env.WDIO_LOG_PATH`](https://github.com/webdriverio/webdriverio/search?q=WDIO_LOG_PATH) always got set in both standalone and runner mode

And also, I tried more with [sequentialWDIOSessions.e2e.js](https://github.com/wyhmichael/wdio-logger-afterSession-issue/blob/main/test/specs/sequentialWDIOSessions.e2e.js#L46), looks like both `process.env.WDIO_LOG_PATH` and `clearLogger` needed in order to switch log path at run time.

So looks like the following would not work, but I am not 100% sure if I use it correctly, any further suggestions?
```
if (!process.env.WDIO_LOG_PATH) {
  logger.clearLogger();
}
```

**Comment by wyhmichael** [comment]
Thanks @christian-bromann ! updated the PR, please take another look 👍

**Comment by christian-bromann** [comment]
> I didn't find any docs to update

I think I will cover this with v8.

Can we extend the unit tests to ensure we cover this code section?

File: packages/webdriver/src/command.ts [hunk 1]
```diff
@@ -101,7 +101,7 @@ export default function (
 
             this.emit('result', { method, endpoint, body, result })
 
-            if (command === 'deleteSession') {
+            if (command === 'deleteSession' && !process.env.KEEP_WDIO_LOGGER) {
```
Comment (christian-bromann) [inline packages/webdriver/src/command.ts:104]: As I mentioned in my post, let's define a `WDIO_WORKER_CID` environment variable within `@wdio/local-runner` and check against its existence here.

File: packages/webdriver/tests/command.test.ts [hunk 2]
```diff
@@ -75,6 +75,31 @@ const scope: BaseClient = new FakeClient()
...
+        expect(method).toBe('POST')
+        expect(endpoint).toBe('/session/:sessionId/element/123/element')
+        expect(using).toBe('css selector')
+        expect(value).toBe('#body')
+        if (additionalParams) {
+            expect(customParam).toBe(additionalParams)
+        }
+        requestMock.mockClear()
+    }
+
```
Comment (christian-bromann) [inline packages/webdriver/tests/command.test.ts:102]: I don't think these tests are relevant to the change you made above, let's revert them.

File: packages/webdriver/tests/command.test.ts [hunk 3]
```diff
@@ -258,18 +282,43 @@ describe('command wrapper result log', () => {
...
+        clearLoggerSpy.mockClear()
+    })
+
+    it('should NOT call clearLogger on deleteSession cmd if env var - KEEP_WDIO_LOGGER set', () => {
+        const resultFunction = getRequestCallback(deleteSessionCmd.method, deleteSessionCmd.path, deleteSessionCmd.endpoint) as unknown as mockResponse
+        process.env.KEEP_WDIO_LOGGER = 'true'
+        resultFunction({})
+        expect((log.info as jest.Mock).mock.calls).toHaveLength(0)
+        expect(clearLoggerSpy).not.toHaveBeenCalled()
+        clearLoggerSpy.mockClear()
```
Comment (christian-bromann) [inline packages/webdriver/tests/command.test.ts:322]: This test si not needed and can be incorporated into `should be no result in log if there is value in response`

File: packages/webdriver/tests/command.test.ts (hunk 2 above)
Comment (wyhmichael) [reply packages/webdriver/tests/command.test.ts]: reverted

File: packages/webdriver/src/command.ts (hunk 1 above)
Comment (wyhmichael) [reply packages/webdriver/src/command.ts]: kk, I am not sure what else the usage of WDIO_WORKER_ID would be, but using it anyways

**Comment by wyhmichael** [comment]
@christian-bromann thanks for the feedback, please take another look at the updated changes

File: packages/webdriver/src/command.ts [hunk 4]
```diff
@@ -101,7 +101,7 @@ export default function (
 
             this.emit('result', { method, endpoint, body, result })
 
-            if (command === 'deleteSession') {
+            if (command === 'deleteSession' && process.env.WDIO_WORKER_ID) {
```
Comment (christian-bromann) [inline packages/webdriver/src/command.ts:104]:
```suggestion
            if (command === 'deleteSession' && !process.env.WDIO_WORKER_ID) {
```

**Review by christian-bromann — COMMENTED** [review]
I think we want this the other way around. Only clear logger when we don't run the wdio testrunner,

File: packages/webdriver/src/command.ts (hunk 4 above)
Comment (wyhmichael) [reply packages/webdriver/src/command.ts]: Updated as suggested, and also updated unit tests

**Review by christian-bromann — APPROVED** [review]
Thanks for the fix 👍

**Comment by christian-bromann** [comment]
It seems that one of the unit tests is not happy:

```
FAIL packages/webdriver/tests/command.test.ts
  ● Test suite failed to run
    Cannot spy the clearLogger property because it is not a function; undefined given instead
      160 |     const log = logger('webdriver')
      161 |     jest.spyOn(log, 'info').mockImplementation((string) => string)
    > 162 |     const clearLoggerSpy = jest.spyOn(logger, 'clearLogger')
          |                                 ^
      163 |
      164 |     function getRequestCallback (method: string, path: string, endpoint: Protocols.CommandEndpoint) {
      165 |         const commandFn = commandWrapper(method, path, endpoint)
      at ModuleMocker.spyOn (node_modules/jest-mock/build/index.js:783:15)
      at packages/webdriver/tests/command.test.ts:162:33
      at Object.<anonymous> (packages/webdriver/tests/command.test.ts:159:1)
```

**Comment by wyhmichael** [comment]
hi @christian-bromann , I wonder if you can help with unit test failure, I have been struggle with it. What I experienced so far: 

if I do: 
1. go under `webdriverio\packages\webdriver`
2. run `tsc .\tests\command.test.ts`
3. run `npx jest -i .\tests\command.test.js` ===> this would work and pass

but I do: 
1. go under `webdriverio`
2. run `npm run test` ===> the unit test would fail for me, I have been trying but haven't got any clue on fixing this yet :(

**Review by christian-bromann — APPROVED** [review]
👍  thanks a lot

**Pull request merged**